### PR TITLE
IGNITE-15964 NPE in case of simultaneous cache destroy and active tx

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java [hunk 1]
```diff
@@ -4569,9 +4569,9 @@ public void awaitLastFut() {
                         ctx.kernalContext().closure().runLocalSafe((GridPlainRunnable)() -> {
                             IgniteInternalFuture fut0;
 
-                            if (ctx.kernalContext().isStopping())
+                            if (ctx.kernalContext().isStopping() || ctx.gate().isStopped())
                                 fut0 = new GridFinishedFuture<>(
-                                    new IgniteCheckedException("Operation has been cancelled (node is stopping)."));
+                                    new IgniteCheckedException("Operation has been cancelled (node or cache is stopping)."));
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java:4574]: Node is stopping, but cache is stopped)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java [hunk 2]
```diff
@@ -191,4 +211,53 @@ private void doTestDestroyCacheOperationNotBlockingCheckpointTest(boolean loc) t
...
+        for (Integer pkey : pkeys)
+            crd.cache(DEFAULT_CACHE_NAME).removeAsync(pkey);
+
+        spi.waitForBlocked();
+
+        spi.blockMessages(GridDhtPartitionsFullMessage.class, getTestIgniteInstanceName(1));
+
+        IgniteInternalFuture destr = GridTestUtils.runAsync(() -> grid(1).destroyCache(DEFAULT_CACHE_NAME));
+
+        spi.waitForBlocked();
```
Comment (vldpyatkov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:253]: You already have blocked messages before, this method always returns true here.
Maybe the block of FullMessage does not require for the test.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java [hunk 3]
```diff
@@ -191,4 +211,53 @@ private void doTestDestroyCacheOperationNotBlockingCheckpointTest(boolean loc) t
...
+                    pkeys.add(i);
+            }
+        }
+
+        TestRecordingCommunicationSpi spi = TestRecordingCommunicationSpi.spi(crd);
+
+        spi.blockMessages(GridDhtTxPrepareRequest.class, getTestIgniteInstanceName(1));
+
+        for (Integer pkey : pkeys)
+            crd.cache(DEFAULT_CACHE_NAME).removeAsync(pkey);
```
Comment (vldpyatkov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:245]: I worry about you do not wait to finish these deletions in the end of the test.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java [hunk 4]
```diff
@@ -191,4 +211,53 @@ private void doTestDestroyCacheOperationNotBlockingCheckpointTest(boolean loc) t
...
     }
+
+    /**
+     * Tests correctness of concurrent cache destroy and implicit tx`s.
+     */
+    @Test
+    public void cacheDestroyWithConcImplicitTx() throws Exception {
+        final IgniteEx crd = (IgniteEx)startGridsMultiThreaded(3);
+
+        crd.cluster().state(ClusterState.ACTIVE);
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:222]: of course, but this test with persistence enabled )

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java (hunk 3 above)
Comment (zstan) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:245]: got it, fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java (hunk 1 above)
Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java:4574]: yep, but i prefer to store it as is.

**Pull request closed without merging**